presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java

```diff
@@ -57,6 +64,34 @@ private ResourceUtil()
 
     public static Session createSessionForRequest(HttpServletRequest servletRequest)
     {
+        String realUser = null;
+        Principal authenticatedPrincipal = servletRequest.getUserPrincipal();
+        if (authenticatedPrincipal != null) {
+            String delegationJsonValue = servletRequest.getHeader(PRESTO_DELEGATION);
+            if (delegationJsonValue != null) {
+                //TODO check if the authenticatedPrincipal has delegation privilege
+                try {
+                    JsonCodec<Delegation> jsonCodec = jsonCodec(Delegation.class);
+                    Delegation delegation = jsonCodec.fromJson(delegationJsonValue);
+                    DelegationIdentity identity = delegation.getIdentity();
+                    if (identity != null) {
+                        if (identity.getType() == UNIX_PRINCIPAL) {
+                            realUser = identity.getName();
+                        }
+                        else if (identity.getType() == KERBEROS_PRINCIPAL) {
+                            realUser = (new KerberosPrincipal(identity.getName())).getName();
+                        }
+                    }
+                }
+                catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException(format("Invalid %s header: %s", PRESTO_DELEGATION, delegationJsonValue));
+                }
+            }
+            else {
+                realUser = authenticatedPrincipal.getName();
+            }
+        }
+
         SessionBuilder sessionBuilder = Session.builder()
                 .setUser(getRequiredHeader(servletRequest, PRESTO_USER, "User"))
```

Comment:
I'm not sure I understand the comment, are you suggesting something like below?
setUser(realUser == null ? getRequiredHeader(servletRequest, PRESTO_USER, "User") : realUser)